L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc

```diff
@@ -95,6 +134,74 @@ void HGCalMulticlusteringImpl::clusterize( const edm::PtrVector<l1t::HGCalCluste
     }
     
 }
+void HGCalMulticlusteringImpl::clusterizeDBSCAN( const edm::PtrVector<l1t::HGCalCluster> & clustersPtrs, 
+                                           l1t::HGCalMulticlusterBxCollection & multiclusters)
+{
 
-
-
+  std::vector<l1t::HGCalMulticluster> multiclustersTmp;
+  l1t::HGCalMulticluster mcluTmp;
+  std::vector<bool> visited(clustersPtrs.size(),false);
+  std::vector<bool> merged (clustersPtrs.size(),false);
+  std::vector<std::vector<int>> neighborList;
+  int iclu = 0, imclu = 0;
+
+  for(edm::PtrVector<l1t::HGCalCluster>::iterator clu = clustersPtrs.begin(); clu != clustersPtrs.end(); ++clu, ++iclu){
+      std::vector<int> neighbors;      
+      
+      if(!visited.at(iclu)){
+	visited.at(iclu)=true;
+	findNeighbor(clustersPtrs, **clu, neighbors);
+	neighborList.push_back(neighbors);
+
+	if(neighborList.at(iclu).size() > minNDbscan_) {
+	  multiclustersTmp.emplace_back( *clu );
+	  merged.at(iclu) = true;
+	  
+	  for(unsigned int neighNo = 0; neighNo < neighborList.at(iclu).size(); neighNo++){
+	    if(!visited.at(neighborList.at(iclu).at(neighNo))){
+	      visited.at(neighborList.at(iclu).at(neighNo)) = true;
+	      std::vector<int> secNeighbors;
+	      findNeighbor(clustersPtrs,*(clustersPtrs[neighborList.at(iclu).at(neighNo)]), secNeighbors);
+	      multiclustersTmp.at(imclu).addConstituent( clustersPtrs[neighborList.at(iclu).at(neighNo)]);
+	      merged.at(neighborList.at(iclu).at(neighNo)) = true;
+	      
+	      if(secNeighbors.size() > minNDbscan_){
+		neighborList.at(iclu).insert(neighborList.at(iclu).end(), secNeighbors.begin(), secNeighbors.end());
+	      }
+	      
+	    } else if(!merged.at(neighborList.at(iclu).at(neighNo)) ){
+	      merged.at(neighborList.at(iclu).at(neighNo)) = true;		
+	      multiclustersTmp.at(imclu).addConstituent( clustersPtrs[neighborList.at(iclu).at(neighNo)] );
+	      }
+	    }
+	    imclu++;
+	}
+      }
+      else neighborList.push_back(neighbors);
```

Comment:
The copy of `neighbors` could be avoided using `neighborList.push_back(std::move(neighbors))`. Though in that case the vector is empty so it's probably not a big difference.